fix: support list in document class (#1557)

maxwelljin opened the pull request:
This pull request resolves the TypeError in BaseDocIndex._flatten_schema that occurs when a non-class type, such as List[str], is used as an argument for issubclass(). As observed in issue #1557, this causes the HnswDocumentIndex initialisation to crash.

As Johannes suggested, I use a custom issubclass function that can handle non-class types by returning False instead of failing. This change prevents crashes when a non-class type is passed as input.

JoanFM reviewed (changes requested):
Please, add tests showing the underlying issue is fixed

maxwelljin commented on docarray/index/abstract.py:
```diff
@@ -776,7 +792,7 @@ def _update_subindex_data(
             cast(Type[BaseDoc], self._schema)
         ):
-            if issubclass(type_, AnyDocArray):
+            if self.custom_issubclass(type_, AnyDocArray):
```
thanks for pointing it out. will fix it

JohannesMessner replied:
Actually I think this should not even be a method, as I say above. That way we can use it in all kinds of places, I don't see a reason why this needs to be tied to BaseDocIndex.

JohannesMessner commented on docarray/index/abstract.py:
```diff
@@ -706,6 +707,21 @@ def _filter_by_parent_id(self, id: str) -> Optional[List[str]]:
     ##########################################################
 
+    @staticmethod
+    def custom_issubclass(x, A_tuple) -> bool:
```
Let's make this a function instead of a staticmethod, and let's more it to `utils._internal._typing`.
Maybe we can also rename it to `safe_issubclass` to indicate a bit more about what makes it custom?

JohannesMessner commented on docarray/index/abstract.py:
```diff
@@ -706,6 +707,21 @@ def _filter_by_parent_id(self, id: str) -> Optional[List[str]]:
+        :return: A boolean value - 'True' if 'x' is a subclass of 'A_tuple', 'False' otherwise.
+                 Note that if the origin of 'x' is a list or tuple, the function immediately returns 'False'.
+        """
+        if get_origin(x) in (list, tuple):
```
What about other origins? I think Dict for example applies in the same way.
Also, are there other things that aren't classes (= break normal issubclass) but also don't have an origin? TypeVars for example? Would be nice to cover as many cases as possible here.

JohannesMessner commented on docarray/index/abstract.py:
```diff
@@ -706,6 +707,21 @@ def _filter_by_parent_id(self, id: str) -> Optional[List[str]]:
+        Traditional 'issubclass' calls can result in a crash if the input is list/tuple.
+
+        :param x: A class 'x'
+        :param A_tuple: A class, or a tuple of classes.
```
let's make this argument lowercase, i.e. `a_tuple`

JohannesMessner commented on docarray/index/abstract.py:
```diff
@@ -706,6 +707,21 @@ def _filter_by_parent_id(self, id: str) -> Optional[List[str]]:
+    @staticmethod
+    def custom_issubclass(x, A_tuple) -> bool:
+        """
+        This is a modified version of the built-in 'issubclass' function to support non-class input.
+        Traditional 'issubclass' calls can result in a crash if the input is list/tuple.
```
Nice docstring!

maxwelljin commented:
Thanks for the feedback! I have made the following changes:

I have added tests (docarray/tests/hnswlib/test_index_get_del.py) to demonstrate that the issue is fixed. These tests verify the functionality and ensure the problem is resolved.
I have converted the method to a function and moved it to utils._internal._typing, as suggested. Additionally, I renamed it to "safe_issubclass" to provide a clearer indication of its purpose.
I have also included support for other origins like "Dict" in addition to "list" and "tuple" and handled cases like "TypeVars" that are not classes but still need to be checked.

Once again, thank you for the valuable feedback

JoanFM commented on docarray/utils/_internal/_typing.py:
```diff
@@ -32,3 +32,17 @@ def change_cls_name(cls: type, new_name: str, scope: Optional[dict] = None) -> N
     cls.__qualname__ = cls.__qualname__[: -len(cls.__name__)] + new_name
     cls.__name__ = new_name
+
+def safe_issubclass(x, a_tuple) -> bool:
```
Please add unittests specific to this method, also add proper type hinting
